code/iaas/cluster-logic/src/main/java/io/cattle/iaas/cluster/service/impl/ClusterManagerImpl.java

```diff
@@ -235,33 +263,397 @@ public boolean evaluate(Instance obj) {
         }
         ConfigUpdateRequest request = ConfigUpdateRequestUtils.getRequest(jsonMapper, state,
                 getContext(clusterAgent));
-        request = before(request, clusterAgent);
+        request = updateClusterConfigBefore(request, clusterAgent);
         ConfigUpdateRequestUtils.setRequest(request, state, getContext(clusterAgent));
-        after(request);
+        updateClusterConfigAfter(request);
     }
 
+    @Override
     public void activateCluster(ProcessState state, Host cluster) {
+        Long currentManagingHostId = DataAccessor.fields(cluster).withKey(ClusterConstants.MANAGING_HOST).as(Long.class);
         Long hostId = findSuitableHost(cluster);
-
-        DataUtils.getWritableFields(cluster).put(ClusterConstants.MANAGING_HOST, hostId);
-        objectManager.persist(cluster);
+        if ((currentManagingHostId == null && hostId != null) ||
+            (currentManagingHostId != null && hostId == null) ||
+            (hostId != null && !hostId.equals(currentManagingHostId))) {
+            DataUtils.getWritableFields(cluster).put(ClusterConstants.MANAGING_HOST, hostId);
+            objectManager.persist(cluster);
+        }
 
         if (hostId == null) {
             processManager.scheduleStandardProcess(StandardProcess.DEACTIVATE, cluster, null);
         } else {
+            try {
+                createClusterCertificates(cluster);
+                createClientCertificates(cluster);
+            } catch (IdempotentRetryException e) {
```

Comment:
Yes, I get why we rethrow idempotentRetry exception here; I'm just not clear why we are catching the rest and wrapping it up in a RuntimeException. As it can potentially convert CheckedExeptions into Runtime ones..